### PR TITLE
Fix calendar emails to be outlook compatible - recreation of #12946

### LINKED ISSUE #12885: Calendar Invites - Outlook integration
### Steps to reproduce
1. Create a Calendar Invite in Thunderbird - invite someone who's using Outlook 2010. 
2. Create a similar Calendar Invite in Nextcloud - invite same person using Outlook 2010

### Expected behaviour
I expect the same behaviour as with Outlook 2016, i.e. the invite gives me buttons on top of the invite and show me the calendar of Outlook/Exchange. However, with the thunderbird invite and invites from google, they are shown; but not with the invites from nextcloud. Thunderbird/nextcloud invites come from the same domain.

image

### Actual behaviour

No buttons are shown. 
image

### Server configuration

**Operating system**: Ubuntu 18

**Web server:** Apache2

**Database:** PostgreSQL 9.5

**PHP version:** 7.2

**Nextcloud version:** (see Nextcloud admin page) 14.x

**Updated from an older Nextcloud/ownCloud or fresh install:** Updated

**Where did you install Nextcloud from:** Web

### PR DESCRIPTION
**Pull request by merlinwoff**
Recreated the abandoned pull request #12946 in order to include review changes. Let's see if that works now ;)

Closes #12885

### CODE REVIEW

File: lib/public/Mail/IMessage.php [hunk 1]
```diff
@@ -40,6 +40,15 @@ interface IMessage {
 	 */
 	public function attach(IAttachment $attachment): IMessage;
 
+	/**
+	 * @param $body:  body of the mime part
+	 * @param $content-type = null: Mime Content-Type (e.g. text/plain or text/calendar)
+	 * @param $charset = null: Character Set (e.g. UTF-8)
+	 * @return IMessage
+	 * @since 24.0.6
```
Comment (ChristophWurst) [inline lib/public/Mail/IMessage.php:48]: needs an update

File: lib/private/Mail/Message.php [hunk 2]
```diff
@@ -63,6 +63,26 @@ public function attach(IAttachment $attachment): IMessage {
...
 	}
 
+	/**
+	 * @param $body:  body of the mime part
+	 * @param $content-type = null: Mime Content-Type (e.g. text/plain or text/calendar)
+	 * @param $charset = null: Character Set (e.g. UTF-8)
+	 * @return $this
+	 * @since 24.0.6
+	 */
+	public function attachInline($data, $contentType = null, $charset = null): IMessage {
```
Comment (tcitworld) [inline lib/private/Mail/Message.php:73]: You can add types for the parameters.

File: lib/public/Mail/IMessage.php [hunk 3]
```diff
@@ -40,6 +40,15 @@ interface IMessage {
...
+	 */
+	public function attachInline($body, $contentType = null, $charset = null): IMessage;
```
Comment (tcitworld) [inline lib/public/Mail/IMessage.php:50]: You can add types for the parameters.

File: lib/private/Mail/Message.php [hunk 4]
```diff
@@ -63,6 +63,26 @@ public function attach(IAttachment $attachment): IMessage {
...
+		# To be sure this works with iCalendar messages, we encode with 8bit instead of
+		# quoted-printable encoding. We save the current encoder, replace the current
+		# encoder with an 8bit encoder and after we've finished, we reset the encoder
+		# to the previous one.
+			$encoder = $this->swiftMessage->getEncoder();
+			$eightbitEncoder = new \Swift_Mime_ContentEncoder_PlainContentEncoder('8bit');
+			$this->swiftMessage->setEncoder($eightbitEncoder);
+			$this->swiftMessage->addPart($data, $contentType, $charset);
```
Comment (tcitworld) [inline lib/private/Mail/Message.php:81]: Does the order of the parts matter? If so and if SwiftMailer doesn't have special any kind of handling on their side, we should add a comment to tell people to use this method *after* the email body already has added.

**Review by tcitworld — COMMENTED** [review]
Nitpicks

File: lib/private/Mail/Message.php (hunk 2 above)
Comment (tcitworld) [reply lib/private/Mail/Message.php:73]: And the first parameter should be named the same than in OCP.

**Review by miaulalala — APPROVED** [review]
LGTM (except for Christoph's comment), the RFC for iMIP explicitly allows this.

Thanks for taking care of this.

File: lib/private/Mail/Message.php (hunk 4 above)
Comment (merlinwoff) [reply lib/private/Mail/Message.php:83]: Unsure if the order matters for swiftmailer (think it doesn't), but I added a warning that maybe not all mail clients play nice if you move it around.

File: lib/private/Mail/Message.php (hunk 2 above)
Comment (merlinwoff) [reply lib/private/Mail/Message.php]: changed the parameter to match, added parameter types

File: lib/public/Mail/IMessage.php (hunk 3 above)
Comment (merlinwoff) [reply lib/public/Mail/IMessage.php]: added parameter types

File: lib/public/Mail/IMessage.php (hunk 1 above)
Comment (merlinwoff) [reply lib/public/Mail/IMessage.php:50]: hopefully it's more clear now?

File: lib/private/Mail/Message.php [hunk 5]
```diff
@@ -63,6 +63,26 @@
...
 	}
 
+	/**
+	 * @param $body:  body of the mime part
+	 * @param $content-type = null: Mime Content-Type (e.g. text/plain or text/calendar)
+	 * @param $charset = null: Character Set (e.g. UTF-8)
+	 * @return $this
+	 * @since 24.0.6
+	 */
+	public function attachInline($data, $contentType = null, $charset = null): IMessage {
```
[the remark this answers is not in the document]
Comment (merlinwoff) [reply lib/private/Mail/Message.php]: fixed that in the meantime

File: lib/public/Mail/IMessage.php (hunk 1 above)
Comment (tcitworld) [reply lib/public/Mail/IMessage.php:50]: It's the `@since` annotation that needs an update. I guess it's too late for 25? So `26.0.0` ?

Comment (merlinwoff) [reply lib/public/Mail/IMessage.php]: @tcitworld ah got it, thanks. Didn't know how your release cycles work really :)

**Review by tcitworld — APPROVED** [review]

File: lib/private/Mail/Message.php [hunk 6]
```diff
@@ -63,6 +63,30 @@
...
+		# To be sure this works with iCalendar messages, we encode with 8bit instead of
+		# quoted-printable encoding. We save the current encoder, replace the current
+		# encoder with an 8bit encoder and after we've finished, we reset the encoder
+		# to the previous one. Originally intended to be added after the message body,
+		# as it is curently unknown if all mail clients handle this properly if added 
+		# before.
+			$encoder = $this->swiftMessage->getEncoder();
+			$eightbitEncoder = new \Swift_Mime_ContentEncoder_PlainContentEncoder('8bit');
+			$this->swiftMessage->setEncoder($eightbitEncoder);
+			$this->swiftMessage->addPart($data, $contentType, $charset);
```
[the remark this answers is not in the document]
Comment (merlinwoff) [reply lib/private/Mail/Message.php]: sorry for that...

**Comment by ChristophWurst** [comment]
The rebase and push broke the PR. Sorry about that.

**Pull request closed without merging**